### ISSUE TITLE
Automatically generate icons for flexible layout templates

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#4104]
Part of the #4046 

In #3755 we have used a single icon for all flexible templates. The idea here is that upon saving a flexible layout template (create/update), the amount of rows and columns/regions per row is calculated, and then these numbers are used to generate an icon/thumbnail to be used the layout template listing in `/admin/structure/layouts/settings`.

---
PR by @klonos: https://github.com/backdrop/backdrop/pull/2915

**Comment by klonos** [comment]
Here's what it looks like:

<img width="955" alt="Screen Shot 2019-10-03 at 10 32 07 am" src="https://user-images.githubusercontent.com/2423362/66091022-3ca2cc00-e5c9-11e9-98f9-df11bd0ce924.png">

Feedback most welcomed 🙂

**Comment by klonos** [comment]
- different color from the standard templates OK?
- should I make the first and last rows narrower than the rest?
- should the vertical/horizontal separators be thicker? (currently 1px)

**Comment by klonos** [comment]
..it does look better with a size of 2px IMO:

<img width="444" alt="Screen Shot 2019-10-03 at 11 22 48 am" src="https://user-images.githubusercontent.com/2423362/66092670-3106d380-e5d0-11e9-9208-390b886defc2.png">

Works fine on my local, but I have pushed changes and the PR sandbox does not comply. Will look into it next chance I get.

**Comment by klonos** [comment]
> - should I make the first and last rows narrower than the rest?

Yup, that looks much better for templates with only a few rows. Compare before/after pairs:

![2019-10-03-15-46-2915 backdrop backdrop qa backdropcms org](https://user-images.githubusercontent.com/2423362/66102308-43473880-e5f5-11e9-863f-59f997c1c907.png)

**Comment by klonos** [comment]
Huh 🤔 ...updating a template on my local also updates it's thumbnail to reflect the new set of rows/regions. This does not work on the PR sandbox though. The thumbnail gets generated initially, then remains the same in future edits.

I am running php 7.2.22 on my local, whereas the PR sandbox is php 5.3.29; so I am assuming that this must have something to do with it. I'll look into it.

**Comment by indigoxela** [comment]
@klonos don't search for problems in code, the actual problem is the browser cache. ;)

The image has the same name and the browser will use the locally cached version until you hit F5 or Apple + r.

**Comment by klonos** [comment]
Huh, I thought that admin pages were not cached. ...but then again these are images, so not pages per se.

Thanks for the heads up @indigoxela 👍 ...care to test/review?

**Comment by klonos** [comment]
...hmm guess another core issue then: https://www.drupal.org/node/2313539 ...we need to allow `theme_image()` to bust cache.

**Comment by indigoxela** [comment]
I've tested your branch on my dev:

Creating and saving works fine. But on /admin/structure/layouts/add, when I try to use the layout for a path, the icon isn't showing.

What I get there is an image tag with an empty _src_ attribute.

![screenshot-icon-missing](https://user-images.githubusercontent.com/842066/66112000-57297500-e5ca-11e9-9cce-fde00a20d6c4.png)

**Comment by klonos** [comment]
Oh sorry, yeah I did not account for that 😞 ...will work on it too. Thanks 👍

**Comment by indigoxela** [comment]
Regarding browser cache: the only thing we could do is to either rename the image on every change (but what about the previous images) or add some "?foo=whatever" to the src attribute, so the browser doesn't reuse the previous version.

**Comment by klonos** [comment]
Yes, I feel that we may need to resort to something like that for this, but in the long-term, I would like us to fix the underlying problem; so here's #4107 😉

**Comment by indigoxela** [comment]
Ah, OK, the referenced D. issue is exactly about that. Didn't get it. ;)

**Comment by quicksketch** [comment]
This looks pretty cool! Nice work sorting out all that GD image work.

Posted a few comments on the PR: https://github.com/backdrop/backdrop/pull/2915#pullrequestreview-327852231

**Comment by klonos** [comment]
I think that this is pretty close to completion. Sorry for not getting back to it sooner.

Setting back to "needs work", as I will need to rebase and address feedback in the PR.

**Comment by klonos** [comment]
I'm planning to loop back to this one once we have #6487 sorted. The current PR is generating .png icons - we'll have to change that to .svg icons instead.

cc @jenlampton (for an overview of the "greater scheme" 🙂)

**Comment by quicksketch** [comment]
@laryn mentioned today during the dev meeting that he has done similar work in Layout Paragraphs that could be ported to Backdrop core. It uses SVGs instead of PNGs, which is probably the direction we want to be headed, especially if https://github.com/backdrop/backdrop-issues/issues/6487 is completed.

**Comment by laryn** [comment]
In the work-in-progress for Paragraphs, I'm generating the SVG markup and then storing it in the flexible layout template config file, rather than trying to save separate files to load individually from the filesystem. I definitely borrowed from @klonos early work above to get this working.

```
/**
 * Generate an SVG image to be used as a Paragraphs layout template thumbnail.
 * @todo If this goes in core it may be possible to use that and remove this.
 * @see https://github.com/backdrop/backdrop-issues/issues/4104
 *
 * @param $template
 *   A layout template object.
 *
 * @return string $svg
 *   The SVG of the generated thumbnail image.
 */
function _layout_paragraphs_template_generate_svg($template) {
  $template_rows = $template->rows;
  $flexi_styles = layout_flexible_row_styles();
  $thumbnail_rows = array();
  $row_color = 'currentColor';
  $svg = "<div class='layout-icon'><svg width='100%' height='100%' viewBox='0 0 100 100' preserveAspectRatio='none'>";
  $thumb_width = 100;
  $thumb_height = 100;
  $separator_size = 3;
  foreach ($template_rows as $row) {
    $thumbnail_rows[] = $flexi_styles[$row['contains']]['split'];
  }
  $rows_count = count($thumbnail_rows);
  $horizontal_separator_count = $rows_count - 1;
  $available_vertical_space = $thumb_height - ($horizontal_separator_count * $separator_size);
  // Divide the vertical space among the number of rows.
  $row_height = ($rows_count) ? $available_vertical_space / ($rows_count) : $available_vertical_space;

  $starting_y = 0;
  foreach ($thumbnail_rows as $index => $region_split) {
    $splits = explode(':', $region_split);
    $regions_count = count(array_filter($splits));
    $vertical_separator_count = $regions_count - 1;
    $available_horizontal_space = $thumb_width - ($vertical_separator_count * $separator_size);
    $current_row_height = $row_height;
    $starting_x = 0;
    foreach ($splits as $split) {
      if (!empty($split) && $starting_x < $thumb_width) {
        $split_width = ($split * $available_horizontal_space) / 100;
        $svg .= "<rect x='$starting_x%' y='$starting_y%' width='" . $split_width . "%' height='" . $current_row_height . "%' fill='$row_color' />";
        $starting_x += $split_width + $separator_size;
      }
    }

    $starting_y += $current_row_height + $separator_size;
  }

  $svg .= "</svg></div>";
  return $svg;
}
```
